### PR TITLE
add nll loss API

### PR DESCRIPTION
**Pull request by joey12300**
### PR types

New features 
### PR changes

APIs
### Describe

Add nll_loss function, and  encapsulate NLL_Loss class and optimize the performance of nll_loss in dygraph mode

### CODE REVIEW

File: python/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -60,3 +66,74 @@
...
+    n = x_shape[0]
+    c = x_shape[1]
+    x_dims = len(x_shape)
+    if x_dims < 2:
+        raise ValueError('Expected 2 or more dimensions (got {})'.format(
+            x_dims))
+    if x_dims != 2 and x_dims != 4:
+        x = paddle.reshape(x, shape=[n, c, 1, -1])
+        label = paddle.reshape(label, shape=[n, 1, -1])
+        out_shape = [n] + x_shape[2:]
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: 这块是否要分一下静态图和动态图？

File: python/paddle/nn/functional/loss.py [hunk 2]
```diff
@@ -60,3 +66,74 @@
...
+
+    if in_dygraph_mode():
+        out, total_weight = core.ops.nll_loss(x, label, weight, 'ignore_index',
+                                              ignore_index, 'reduction',
+                                              reduction)
+        if x_dims != 2 and x_dims != 4 and reduction == 'none':
+            out = paddle.reshape(out, shape=out_shape)
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: 用core.ops

File: python/paddle/nn/layer/loss.py [hunk 3]
```diff
@@ -505,18 +506,18 @@ class NLLLoss(fluid.dygraph.Layer):
         \\end{cases}
 
     Parameters:
-        input (Variable): Input tensor, the data type is float32, float64. 
+        x (Variable): Input tensor, the data type is float32, float64.
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py]: Variable -> Tensor  下面都是把Variable改成Tensor

File: python/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -60,3 +66,74 @@
...
+    Returns:
+         The tensor variable storing the nll_loss.
+  
+    Return type: Variable.
+    """
+    if reduction not in ['sum', 'mean', 'none']:
+        raise ValueError(
+            "The value of 'reduction' in nll_loss should be 'sum', 'mean' or "
+            "'none', but received %s, which is not allowed." % reduction)
+    x_shape = list(x.shape)
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: 要确认一下Tensor这种shape属性的调用方式

File: python/paddle/nn/functional/loss.py [hunk 5]
```diff
@@ -60,3 +66,123 @@
...
+
+
+def nll_loss(x, label, weight=None, ignore_index=-100, reduction='mean'):
+    """
+    This api returns negative log likelihood.
+    See more detail in :ref:`api_nn_loss_NLLLoss` .
+
+    Parameters:
+         x (Tensor): Input tensor, the data type is float32, float64.
+         label (Tensor): Label tensor, the data type is int64_t.
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: Int64_t -> int64

Comment (wawltor) [reply python/paddle/nn/functional/loss.py]: 按照规范copy一下

File: python/paddle/nn/functional/loss.py [hunk 6]
```diff
@@ -60,3 +66,123 @@
...
+         weight (Tensor, optional): Weight tensor, a manual rescaling weight given
+             to each class. If given, it has to be a Tensor of size `C`. Otherwise,
+             it treated as if having all ones. the data type is
+             float32, float64, Default is ``'None'``.
+         ignore_index (int64, optional): Specifies a target value that is ignored
+             and does not contribute to the input gradient.
+         reduction (str, optional): Indicate how to average the loss,
+             the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
+             If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: sum也要介绍一下

File: python/paddle/nn/functional/loss.py [hunk 7]
```diff
@@ -60,3 +66,123 @@
...
+                    label = paddle.nn.data(name='label', shape=[10], dtype='int64')
+                    log_out = log_softmax(x)
+                    res = nll_loss(log_out, label)
+
+                    exe = paddle.Executor(place)
+                    declaritive_result = exe.run(
+                        prog,
+                        feed={"x": x_np,
+                              "label": label_np},
+                        fetch_list=[res])
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: 静态图示例可以去掉

File: python/paddle/nn/functional/loss.py [hunk 8]
```diff
@@ -60,3 +66,123 @@
...
+            x_dims))
+    n = x_shape[0]
+    c = x_shape[1]
+
+    if x_dims != 2 and x_dims != 4:
+        x = paddle.reshape(x, shape=[n, c, 1, -1])
+        label = paddle.reshape(label, shape=[n, 1, -1])
+        out_shape = [n] + x_shape[2:]
+
+    helper = LayerHelper('nll_loss', **locals())
```
Comment (wawltor) [inline python/paddle/nn/functional/loss.py]: 这块提前一下

File: python/paddle/nn/layer/loss.py [hunk 9]
```diff
@@ -505,106 +506,73 @@ class NLLLoss(fluid.dygraph.Layer):
...
             it treated as if having all ones. the data type is 
             float32, float64, Default is ``'None'``.
+        ignore_index (int64, optional): Specifies a target value that is ignored
+            and does not contribute to the input gradient.
         reduction (str, optional): Indicate how to average the loss, 
             the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
             If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned; 
             Default is ``'mean'``.
-        ignore_index (int64, optional): Specifies a target value that is ignored
-            and does not contribute to the input gradient.
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py:508]: 增加name参数

File: python/paddle/nn/layer/loss.py [hunk 10]
```diff
@@ -505,106 +506,73 @@ class NLLLoss(fluid.dygraph.Layer):
...
 
     Returns:
-        The tensor variable storing the nll_loss.
+        The callable object which calculates negative log likelihood loss when 
+        get the input `x` and `label` .
```
Comment (wawltor) [inline python/paddle/nn/layer/loss.py]: 去掉x和label相关信息

**Review by wawltor — DISMISSED** [review]

File: python/paddle/nn/functional/loss.py [hunk 11]
```diff
@@ -60,3 +67,117 @@
...
+         label (Tensor): Label tensor, the data type is int64.
+         weight (Tensor, optional): Weight tensor, a manual rescaling weight given
+             to each class. If given, it has to be a Tensor of size `C`. Otherwise,
+             it treated as if having all ones. the data type is
+             float32, float64, Default is ``'None'``.
+         ignore_index (int64, optional): Specifies a target value that is ignored
+             and does not contribute to the input gradient.
+         reduction (str, optional): Indicate how to average the loss,
+             the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
+             If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
```
Comment (TCChenlong) [inline python/paddle/nn/functional/loss.py]: 格式应该有问题 【If `reduction` is】就可以 下面同理

**Review by wawltor — DISMISSED** [review]
LGTM

**Review by yaoxuefeng6 — DISMISSED** [review]
LGTM

File: python/paddle/nn/functional/loss.py [hunk 12]
```diff
@@ -60,3 +67,115 @@
...
+                result = nll_loss(log_out, label)
+                print(result.numpy()) # [1.0720209]
+    """
+    if reduction not in ['sum', 'mean', 'none']:
+        raise ValueError(
+            "The value of 'reduction' in nll_loss should be 'sum', 'mean' or "
+            "'none', but received %s, which is not allowed." % reduction)
+
+    if in_dygraph_mode():
+        x_shape = list(core.ops.shape(x))
```
Comment (zhiqiu) [inline python/paddle/nn/functional/loss.py]: you can write `x_shape = x.shape` directly here

File: python/paddle/nn/layer/loss.py [hunk 13]
```diff
@@ -505,106 +506,72 @@ class NLLLoss(fluid.dygraph.Layer):
...
-        outputs = {'Out': out, 'Total_weight': total_weight}
-
-        self._helper.append_op(
-            type='nll_loss', inputs=inputs, outputs=outputs, attrs=attrs)
-        if x_dims != 2 and x_dims != 4 and self.reduction == 'none':
-            out = fluid.layers.reshape(out, shape=out_shape)
+                "The value of 'reduction' in nll_loss should be 'sum', 'mean' or "
+                "'none', but received %s, which is not allowed." % reduction)
+        super(NLLLoss, self).__init__()
+        self.weight = weight
```
Comment (zhiqiu) [inline python/paddle/nn/layer/loss.py]: maybe `self._weight` is better for private member, etc.

Comment (joey12300) [reply python/paddle/nn/layer/loss.py]: done

File: python/paddle/nn/functional/loss.py (hunk 12 above)
Comment (joey12300) [reply python/paddle/nn/functional/loss.py]: done

File: python/paddle/nn/functional/loss.py [hunk 14]
```diff
@@ -60,3 +67,108 @@
...
+             name=None):
+    """
+    This api returns negative log likelihood.
+    See more detail in :ref:`api_nn_loss_NLLLoss` .
+
+    Parameters:
+         x (Tensor): Input tensor, the data type is float32, float64.
+         label (Tensor): Label tensor, the data type is int64.
+         weight (Tensor, optional): Weight tensor, a manual rescaling weight given
+             to each class. If given, it has to be a Tensor of size `C`. Otherwise,
```
Comment (zhiqiu) [inline python/paddle/nn/functional/loss.py]: > it has to be a Tensor of size `C`

Better describe this clearly, does it means x should be of format [N, C, H, W]

Comment (joey12300) [reply python/paddle/nn/functional/loss.py]: > > it has to be a Tensor of size `C`
> 
> Better describe this clearly, does it means x should be of format [N, C, H, W]

 @zhiqiu Already describe the shape of `x`, `label`, `weight` more clearly.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by wawltor — APPROVED** [review]
LGTM

**Pull request merged**